plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2055,7 +2069,47 @@ void producer_plugin_impl::produce_block() {
       return sigs;
    } );
 
-   chain.commit_block();
+   // Notes:
+   // - possibly use `std::future` in the future:
+   //   - caller that gets the future can decide thread details.
+   //   - look into thread pooling.
+   // - `propose_constructed_block` can fail in it's current implementation.
+   // - `append_external_block` cannot fail in it's current implementation.
+   // - to get the watermark:
+   //   - call `get_watermark`.
+   // - to get the producer:
+   //   - call `hbs->get_scheduled_producer(block_time)`.
+   //   - or call `chain.head_block_state()` of which you might call `account_name controller::pending_block_producer() const`.
+   // - to get LIBid or LIBnumber:
+   //   - both can be found in controller.
+   //   - for this case do not get the chain last irreversible block number but get the LIB number from the head block state.
+   //   - get dpos irreversible block number
+   //   - block header state also has the producer.
+   //   - block header state also has the signed block as well.
+   // - it was talked about how another function could be added to controller for dealing with precursor cases prior to calling `commit_block`.
+   // - not sure why I wrote this function down but I'm keeping it around: `chain.fetch_block_by_number()`.
+   // - also keeping this note around `bsp = chain.finalize_block( [&]( const digest_type& d )` // get dpos irreversible block number
+
+   if ( blockvault_plug ) {
+      blockvault::watermark_t watermark{get_watermark(chain.head_block_state()->get_scheduled_producer(calculate_pending_block_time()).producer_name).value()};
```

Comment:
pending block is the block state you want, not `head_block_state`.
`finialize_block` returns the `block_state_ptr`, use that.
get_watermark can return an empty optional. That should not be possible in this case, but best to EOS_ASSERT that.
Use the `producer` of the block, not the scheduled producer.

Reply:
Changed code to use the `block_state_ptr` (pending block) returned from `finalize_block`; changed code to check for the empty optional case; changed code to use the producer of the block instead of the scheduled producer.